File: scripts/archive/dentalink-deprecated/poll-and-download.mjs
/**
 * Poll estado and download completed reports
 */
import { chromium } from 'playwright'
import { writeFileSync, mkdirSync, readFileSync } from 'fs'
import { join, dirname } from 'path'
import { fileURLToPath } from 'url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const ROOT = join(__dirname, '..')
const OUT_DIR = join(ROOT, 'data', 'imports', 'pagos')
mkdirSync(OUT_DIR, { recursive: true })

const [,, phpsessid] = process.argv
const PHPSESSID = phpsessid || process.env.DENTALINK_PHPSESSID
if (!PHPSESSID) { console.error('Need PHPSESSID'); process.exit(1) }

// Poll settings
const POLL_INTERVAL_MS = 30000  // 30 seconds
const MAX_POLLS = 40  // up to 20 minutes total wait

async function checkSolicitudes(phpsessid) {
  const response = await fetch(`https://drdiente.dentalink.cl/solicitudes_reportes/historial`, {
    headers: { 'Cookie': `PHPSESSID=${phpsessid}` }
  })
  const html = await response.text()
  
  // Parse the dataSolicitudes from HTML
  const match = html.match(/window\.dataSolicitudes\s*=\s*(\[[\s\S]*?\]);/)
  if (!match) return []
  return JSON.parse(match[1])
}

async function downloadReport(solicitud) {
  const id = solicitud.id
  const nombre = solicitud.nombre
  const s3path = solicitud.s3_path
  const bucket = solicitud.s3_bucket
  const estado = solicitud.id_estado
  const options = JSON.parse(solicitud.options || '{}')
  
  console.log(`ID ${id}: ${nombre} (estado ${estado}, s3_path=${s3path || 'none'})`)
  console.log(`  Options: ${JSON.stringify(options).slice(0, 100)}`)
  
  if (estado !== 3 || !s3path) return false
  
  // Generate output filename
  let outName
  if (nombre === 'finanzas.pagos') outName = 'pagos-globales'
  else if (nombre === 'finanzas.pagos_detalle_accion') outName = 'pagos-acciones-servicio'
  else outName = nombre.replace(/\./g, '-')
  
  // Try direct S3 download first
  if (s3path && bucket) {
    const s3Url = `https://${bucket}.s3.amazonaws.com/${s3path}`
    console.log(`  Downloading from S3: ${s3Url}`)
    
    const response = await fetch(s3Url)
    if (response.ok) {
      const buffer = await response.arrayBuffer()
      const filePath = join(OUT_DIR, `${outName}-full.xlsx`)
      writeFileSync(filePath, Buffer.from(buffer))
      console.log(`  ✅ Saved: ${filePath}`)
      return true
    } else {
      console.log(`  S3 direct failed (${response.status}), trying via browser...`)
    }
  }
  
  // Fallback: use Playwright to click the download button in historial
  console.log(`  Using Playwright to download ID ${id}...`)
  const browser = await chromium.launch({
    headless: true, args: ['--no-sandbox', '--disable-dev-shm-usage'],
  })
  try {
    const context = await browser.newContext({
      locale: 'es-MX', viewport: { width: 1440, height: 900 }, acceptDownloads: true,
    })
    await context.addCookies([{
      name: 'PHPSESSID', value: phpsessid.trim(),
      domain: 'drdiente.dentalink.cl', path: '/', httpOnly: true, secure: true, sameSite: 'None',
    }])
    const page = await context.newPage()
    await page.goto('https://drdiente.dentalink.cl/solicitudes_reportes/historial', {
      waitUntil: 'domcontentloaded', timeout: 60000
    })
    await page.waitForTimeout(3000)
    
    // Find the row with this solicitud ID and click Descargar
    const rows = page.locator('table tr, div[class*="row"]')
    const count = await rows.count()
    for (let i = 0; i < count; i++) {
      const text = await rows.nth(i).innerText()
      if (text.includes(` ${id}\t`) || text.includes(`\n${id}\n`)) {
        console.log(`  Found row ${i} for ID ${id}: "${text.slice(0, 60)}"`)
        const downloadBtn = rows.nth(i).locator('button, a').filter({ hasText: /Descargar/i })
        if (await downloadBtn.count()) {
          const dlPromise = page.waitForEvent('download', { timeout: 30000 })
          await downloadBtn.first().click()
          await page.waitForTimeout(2000)
          const dl = await dlPromise
          const filePath = join(OUT_DIR, `${outName}-full.xlsx`)
          await dl.saveAs(filePath)
          console.log(`  ✅ Downloaded via Playwright: ${filePath}`)
          return true
        }
        break
      }
    }
  } finally {
    await browser.close()
  }
  return false
}

console.log('=== Polling Dentalink reports ===')
console.log(`Looking for reports 168 (finanzas.pagos) and 169 (finanzas.pagos_detalle_accion)`)

let found168 = false, found169 = false

for (let poll = 1; poll <= MAX_POLLS; poll++) {
  console.log(`\n--- Poll ${poll} ---`)
  
  const solicitudes = await checkSolicitudes(PHPSESSID)
  console.log(`Total solicitudes on page: ${solicitudes.length}`)
  
  for (const s of solicitudes) {
    const id = parseInt(s.id)
    if (id === 168) {
      console.log('  ID 168 (finanzas.pagos):')
      found168 = await downloadReport(s) || found168
    }
    if (id === 169) {
      console.log('  ID 169 (finanzas.pagos_detalle_accion):')
      found169 = await downloadReport(s) || found169
    }
  }
  
  if (found168 && found169) {
    console.log('\n✅ Both reports downloaded!')
    process.exit(0)
  }
  
  if (poll < MAX_POLLS) {
    console.log(`\nWaiting ${POLL_INTERVAL_MS/1000}s before next poll...`)
    await new Promise(r => setTimeout(r, POLL_INTERVAL_MS))
  }
}

console.log('Max polls reached. Reports may still be generating.')
process.exit(1)

File: scripts/archive/dentalink-deprecated/poll-and-download.py
#!/usr/bin/env python3
"""Poll historial for report IDs 168 and 169 and download via S3 when ready."""
import urllib.request, json, re, time, os, sys

PHPSESSID = "7tk4mf24dli8ghhk57uthbv8tkg177fc"
BASE = "https://drdiente.dentalink.cl"
OUT_DIR = "/home/hermes/leads-admin/data/imports/pagos"
os.makedirs(OUT_DIR, exist_ok=True)

# New IDs with correct params: sucursal_id="all", documentos="all"
TARGETS = {"170": "pagos-globales", "171": "pagos-acciones-servicio"}

def check():
    req = urllib.request.Request(f"{BASE}/solicitudes_reportes/historial")
    req.add_header("Cookie", f"PHPSESSID={PHPSESSID}")
    resp = urllib.request.urlopen(req, timeout=30)
    html = resp.read().decode()
    # Use regex to find the big JSON array with all reports
    import re
    matches = list(re.finditer(r'\[\{"id".*?}\]', html, re.DOTALL))
    for m in matches:
        data = json.loads(m.group())
        if len(data) < 50:
            continue
        results = {}
        for item in data:
            sid = item['id']
            if sid in TARGETS:
                results[sid] = item
        return results, None
    return None, "No report data found"

import sys
print(f"Polling for reports {list(TARGETS.keys())}...")
for i in range(60):
    results, err = check()
    if err:
        print(f"  [{i+1}] Error: {err}")
        time.sleep(30)
        continue
    
    ready = True
    for sid, name in TARGETS.items():
        if sid not in results:
            print(f"  [{i+1}] ID {sid} not yet in historial")
            ready = False
            continue
        
        item = results[sid]
        estado = item.get('id_estado', '?')
        s3 = item.get('s3_path', '')
        status_str = f"estado={estado}"
        if s3:
            status_str += f" s3_path={s3[:40]}..."
        print(f"  [{i+1}] ID {sid} ({name}): {status_str}")
        
        if estado == '3' and s3:
            # Download!
            s3_url = f"https://{item['s3_bucket']}.s3.amazonaws.com/{s3}"
            out_name = f"{name}-full.xlsx"
            out_path = os.path.join(OUT_DIR, out_name)
            
            req2 = urllib.request.Request(s3_url)
            resp2 = urllib.request.urlopen(req2, timeout=120)
            data2 = resp2.read()
            with open(out_path, 'wb') as f:
                f.write(data2)
            mb = len(data2) / (1024*1024)
            print(f"    ✅ DOWNLOADED: {out_path} ({mb:.1f} MB)")
        else:
            ready = False
    
    if ready:
        print("\n✅ AMBOS REPORTES DESCARGADOS!")
        sys.exit(0)
    
    print(f"  Waiting 30s...\n")
    time.sleep(30)

print("Timeout - reports not ready yet")
